File: J9Ch/src/J_2_Binary_Search/Required_10/python_cpp/_447_Search_in_a_Big_Sorted_Array_Medium.py
'''/**447. Search in a Big Sorted Array
 * Medium'''


"""
Definition of ArrayReader:
class ArrayReader:
    def get(self, index):
        # this would return the number on the given index
        # return -1 if index is less than zero.
"""
class ArrayReader:
    def get(self, index):
        return 1;

class Solution:
    # @param {ArrayReader} reader: An instance of ArrayReader
    # @param {int} target an integer
    # @return {int} an integer
    def searchBigSortedArray(self, reader, target):
        index = 0
        while reader.get(index) < target:
            index = index * 2 + 1

        start, end = 0, index
        while start + 1 < end:
            mid = (start + end) / 2
            if reader.get(mid) >= target:
                end = mid
            else:
                start = mid

        if reader.get(start) == target:
            return start
        if reader.get(end) == target:
            return end
        return -1

File: J9Ch/src/J_2_Binary_Search/Required_10/python_cpp/_447_Search_in_a_Big_Sorted_Array_Medium.cpp
/**447. Search in a Big Sorted Array
 * Medium*/

/**
 * Definition of ArrayReader:
 *
 * class ArrayReader {
 * public:
 *     int get(int index) {
 *          // return the number on given index,
 *          // return -1 if index is less than zero.
 *     }
 * };
 */
class ArrayReader {
public:
    int get(int index) {
        return 1;
    }
}

class Solution {
public:
    /**
     * @param reader: An instance of ArrayReader.
     * @param target: An integer
     * @return: An integer which is the first index of target.
     */
    int searchBigSortedArray(ArrayReader *reader, int target) {
        int e = 1;
        while (reader->get(e) < target) {
            e *= 2;
        }

        int s = 0;
        while (s + 1 < e) {
            int mid = s + (e - s) / 2;
            int v = reader->get(mid);
            if (v == target) {
                e = mid;
            } else if (v < target) {
                s = mid;
            } else {
                e = mid;
            }
        }
        if (reader->get(s) == target) {
            return s;
        } else if (reader->get(e) == target) {
            return e;
        }
        return -1;
    }
};